Get sha of the image after Container.From

LokeshN opened the pull request:
References https://github.com/dagger/dagger/issues/4516
The following PR contains the proposal for https://github.com/dagger/dagger/issues/4516
- introduced imageRef field to the container payload
- exposed the imageRef in the api
- logic is to get the image ref in From() and set it to the container payload and if there are any modifications to the container with* methods, then make it empty and when it is empty, throw an error as suggested in the issue.
- introduced a new method containerFromPayload(payload) as it can be reused in multiple places.
- Generated sdk for node, go and python
**Tested the following scenarios:**
I/P: 

```
        {
               container {
			from(address:"alpine-latest") {
				imageRef
			}
		}
	}
```

O/P:

```
 {
    "container": {
        "from": {
            "imageRef": "7cd52847ad775a5ddc4b58326cf884beee34544296402c6292ed76474c686d39"
        }
    }
}
```

I/P:

```
        {
		container {
			from(address:"hello-world") {
				exec(args:["/hello"]) {
				  imageRef
				}
			  }
			
		}
	}
```

O/P:

`Error: failed to solve: Image reference can only be retrieved immediately after the 'Container.From' call. Error in fetching imageRef as the container image is changed`

LokeshN commented:
@sipsma if you get some time, can you review this PR. Thanks.

jlongtine commented:
> @sipsma if you get some time, can you review this PR. Thanks.

@LokeshN As an FYI, @sipsma is on vacation this week. 

@vito / @aluzzardi: Could one of you take a look?

vito commented on core/container.go:
```diff
@@ -258,6 +261,25 @@ func (container *Container) From(ctx context.Context, gw bkgw.Client, addr strin
+	}
+
+	fs := imgSpec.RootFS
+	if len(fs.DiffIDs) > 0 {
```
I think the digest you're looking for is actually returned already as the first discarded value on this line:

```go
_, cfgBytes, err := gw.ResolveImageConfig(ctx, ref, llb.ResolveImageConfigOpt{
```

I think you could tweak it like this:

```go
	ref := reference.TagNameOnly(refName)

	digest, cfgBytes, err := gw.ResolveImageConfig(ctx, ref.String(), llb.ResolveImageConfigOpt{
		// ...
	})
	if err != nil {
		return nil, err
	}

	digested, err := reference.WithDigest(ref, digest)
	if err != nil {
		return nil, err
	}
```

and then store `digested.String()` as the `ImageRef`.

LokeshN replied:
fixed

vito commented on core/container.go:
```diff
@@ -1322,6 +1312,29 @@ func (container *Container) export(
 	return res, nil
 }
 
+func (container *Container) Sha(ctx context.Context, gw bkgw.Client) (string, error) {
```
IMO `Sha` is pretty unclear. I think it would be better to just call it `ImageRef` and have it return a ref like `my/repo@sha256:...`.

LokeshN replied:
fixed

vito commented on core/schema/container.graphqls:
```diff
@@ -297,6 +297,9 @@ type Container {
   ): Container!
+
+  "The unique image reference"
+  sha: String
```
Same here, I would just call it `imageRef`.

LokeshN replied:
fixed

vito reviewed (commented):
Thanks for starting on this, and sorry for the wait! Left some feedback inline. We'll also need a test for this, which can be added to `core/integration/container_test.go`.

It looks like our contributing docs could use some updates, so here's a quick primer:

It looks like you already figured out how to regenerate the SDK, but for posterity it's `./hack/make sdk:all:generate`.

This command will build and run a local engine with your latest changes and run the integration tests:

```sh
./hack/dev go test ./core/integration
```

There are a lot of tests, so you might want to focus yours:

```sh
./hack/dev go test -v ./core/integration -run TestFooBar
```

LokeshN commented:
> > @sipsma if you get some time, can you review this PR. Thanks.
> 
> @LokeshN As an FYI, @sipsma is on vacation this week.
> 
> @vito / @aluzzardi: Could one of you take a look?

Thanks @jlongtine for the help!

LokeshN commented:
> Thanks for starting on this, and sorry for the wait! Left some feedback inline. We'll also need a test for this, which can be added to `core/integration/container_test.go`.
> 
> It looks like our contributing docs could use some updates, so here's a quick primer:
> 
> It looks like you already figured out how to regenerate the SDK, but for posterity it's `./hack/make sdk:all:generate`.
> 
> This command will build and run a local engine with your latest changes and run the integration tests:
> 
> ```shell
> ./hack/dev go test ./core/integration
> ```
> 
> There are a lot of tests, so you might want to focus yours:
> 
> ```shell
> ./hack/dev go test -v ./core/integration -run TestFooBar
> ```

Thanks @vito 🙏 . Fixed the issues and added tests.

sipsma commented on core/schema/container.graphqls:
```diff
@@ -557,6 +557,9 @@ type Container {
     address: String!
   ): Container!
+
+  "The unique image reference"
```
I think we should mention the caveats about this currently only working right after a `Container.From` in this doc string too.

LokeshN replied:
Thanks Eric @sipsma, fixed it.

sipsma reviewed (commented):
One easy fix comment, otherwise LGTM, thanks again!

vito reviewed (approved):
Looks great, thanks again!
